plutus-emulator/src/Wallet/Emulator/Types.hs

```diff
@@ -504,13 +506,22 @@ liftMockWallet wallet act = do
     let walletState = fromMaybe (emptyWalletState wallet) $ Map.lookup wallet $ _walletStates emState
         ((out, newState), (msgs, txns)) = runWriter $ runStateT (runExceptT (runMockWallet act)) walletState
         events = (TxnSubmit . hashTx <$> txns) ++ (WalletInfo wallet <$> getWalletLog msgs)
-    put emState {
-        _txPool = txns ++ _txPool emState,
-        _walletStates = Map.insert wallet newState $ _walletStates emState,
-        _emulatorLog = events ++ _emulatorLog emState
-        }
+    liftNodeClient $ mapM_ NC.publishTx txns
+    modify $ set walletStates (Map.insert wallet newState $ _walletStates emState)
+           . set emulatorLog  (events ++ _emulatorLog emState)
     pure (txns, out)
 
+liftNodeClient :: MonadState EmulatorState m => State NC.NodeClientState a -> m a
+liftNodeClient ncAction = state $ \st ->
+    let (v, newNcState) = runState ncAction (_nodeClientState st)
+    in  (v, st & nodeClientState .~ newNcState)
+
+fromNodeEvent :: NC.NodeClientEvent -> EmulatorEvent
+fromNodeEvent (NC.TxnSubmit tid) = TxnSubmit tid
+fromNodeEvent (NC.TxnValidate tid) = TxnValidate tid
+fromNodeEvent (NC.TxnValidationFail tid err) = TxnValidationFail tid err
+fromNodeEvent (NC.SlotAdd newSlot) = SlotAdd newSlot
```

Comment:
Seem to me like we should just add a constructor that wraps `NodeClientEvent` rather than duplicating them all.